{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import itertools\n",
    "from scipy import stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_gt(filename):\n",
    "    with open(filename, \"r\") as fileHandler:\n",
    "        lines = fileHandler.readlines()\n",
    "    return lines\n",
    "\n",
    "def get_pop(filename):\n",
    "    phenotypes = []\n",
    "    ids = []\n",
    "    with open(filename, \"r\") as fileHandler:\n",
    "        lines = fileHandler.readlines()\n",
    "    for i in range(len(lines)):\n",
    "        ids.append(lines[i].split()[0])\n",
    "        phenotypes.append(lines[i].split()[2])\n",
    "    return ids, phenotypes\n",
    "\n",
    "def get_snp_ids(filename):\n",
    "    snp_ids = ['IND', 'X', 'X', 'X', 'X', 'Pheno']\n",
    "    with open(filename, \"r\") as fileHandler:\n",
    "        lines = fileHandler.readlines()\n",
    "    for i in range(len(lines)):\n",
    "        snp_ids.append(lines[i].split()[1])\n",
    "        snp_ids.append(lines[i].split()[1])\n",
    "    return snp_ids\n",
    "\n",
    "def get_ids(genotypes):\n",
    "    ids = []\n",
    "    for i in range(0, len(genotypes)):\n",
    "        ids.append(genotypes[i].split()[0])\n",
    "    return ids\n",
    "\n",
    "def get_data_frame(genotypes, snp_ids):\n",
    "    inds = []\n",
    "    for i in range(0, len(genotypes)):\n",
    "        inds.append(genotypes[i].split())\n",
    "    df = pd.DataFrame(inds)\n",
    "    df.columns = snp_ids\n",
    "    return df\n",
    "\n",
    "def which_minor_al(frq1, frq2, a1, a2):\n",
    "    if frq1 < frq2:\n",
    "        return frq1, a1\n",
    "    else:\n",
    "        return frq2, a2\n",
    "\n",
    "def get_al_freq(snp_id, pop):\n",
    "    aa = 0\n",
    "    bb = 0\n",
    "    ab = 0\n",
    "    total = 2*len(pop[[snp_id]])\n",
    "    al_1 = \"1\"\n",
    "    al_2 = \"2\"\n",
    "\n",
    "    for i in range(len(pop[[snp_id]])):\n",
    "        \n",
    "        if pop[[snp_id]].iloc[i].iloc[0] == '2' and pop[[snp_id]].iloc[i].iloc[1] == '2':\n",
    "            aa += 1\n",
    "        elif pop[[snp_id]].iloc[i].iloc[0] == '1' and pop[[snp_id]].iloc[i].iloc[1] == '1':\n",
    "            bb += 1\n",
    "        else:\n",
    "            ab += 1\n",
    "\n",
    "    obs_a = 2*aa + ab\n",
    "    obs_b = 2*bb + ab\n",
    "\n",
    "    f1 = (2*aa + ab) / total\n",
    "    f2 = (2*bb + ab) / total\n",
    "\n",
    "    return which_minor_al(f1, f2, al_1, al_2)\n",
    "\n",
    "def get_missing(snp_id, df):\n",
    "    missing_nu = 0\n",
    "    for i in range(len(df[[snp_id]])):\n",
    "        #if df[[snp_id]].iloc[i].all() == '0':\n",
    "         #   missing_nu += 2\n",
    "        if df[[snp_id]].iloc[i].any() == '0':\n",
    "            missing_nu += 1\n",
    "    return missing_nu, missing_nu / len(df[[snp_id]])\n",
    "\n",
    "def missing_snps(person, nu_snps):\n",
    "    missing = 0\n",
    "    for i in range(nu_snps-6):\n",
    "        print(i+6)\n",
    "        if person[i+6] == '0' or person[i+7] == '0':\n",
    "            missing += 1\n",
    "        i += 2\n",
    "    return missing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def chi_square_test(pop):\n",
    "    aa_aff = 0\n",
    "    bb_aff = 0\n",
    "    ab_aff = 0\n",
    "\n",
    "    aa_unaf = 0\n",
    "    bb_unaf = 0\n",
    "    ab_unaf = 0\n",
    "\n",
    "    AFFECTED = '2'\n",
    "    CONTROL = '1'\n",
    "\n",
    "    for i in range(len(pop)):\n",
    "        if pop.iloc[i].iloc[1] == '2' and pop.iloc[i].iloc[2] == '2' and pop.iloc[i].iloc[0] == AFFECTED:\n",
    "            aa_aff += 1\n",
    "        elif pop.iloc[i].iloc[1] == '1' and pop.iloc[i].iloc[2] == '1' and pop.iloc[i].iloc[0] == AFFECTED:\n",
    "            bb_aff += 1\n",
    "        elif pop.iloc[i].iloc[0] == AFFECTED:\n",
    "            ab_aff += 1\n",
    "        \n",
    "        if pop.iloc[i].iloc[1] == '2' and pop.iloc[i].iloc[2] == '2' and pop.iloc[i].iloc[0] == CONTROL:\n",
    "            aa_unaf += 1\n",
    "        elif pop.iloc[i].iloc[1] == '1' and pop.iloc[i].iloc[2] == '1' and pop.iloc[i].iloc[0] == CONTROL:\n",
    "            bb_unaf += 1\n",
    "        elif pop.iloc[i].iloc[0] == CONTROL:\n",
    "            ab_unaf += 1\n",
    "\n",
    "    a = 2*aa_aff + ab_aff\n",
    "    b = 2*bb_aff + ab_aff\n",
    "    c = 2*aa_unaf + ab_unaf\n",
    "    d = 2*bb_unaf + ab_unaf\n",
    "\n",
    "    total = a + b + c + d\n",
    "\n",
    "    #print (a, b, c, d)\n",
    "    a_aff = a / total\n",
    "    b_aff = b / total\n",
    "    a_unaf = c / total\n",
    "    b_unaf = d / total\n",
    "    #print(a_aff, b_aff, a_unaf, b_unaf)\n",
    "\n",
    "    p_aff = a_aff + b_aff\n",
    "    p_unaf = a_unaf + b_unaf\n",
    "    p_a = a_aff + a_unaf\n",
    "    p_b = b_aff + b_unaf\n",
    "    #print(p_aff, p_unaf, p_a, p_b)\n",
    "\n",
    "    p_a_aff = p_aff * p_a\n",
    "    p_b_aff = p_aff * p_b\n",
    "    p_a_unaf = p_unaf * p_a\n",
    "    p_b_unaf = p_unaf * p_b\n",
    "    #print(p_a_aff, p_b_aff, p_a_unaf, p_b_unaf)\n",
    "\n",
    "    exp_a_aff = total * p_a_aff\n",
    "    exp_b_aff = total * p_b_aff\n",
    "    exp_a_unaf = total * p_a_unaf\n",
    "    exp_b_unaf = total * p_b_unaf\n",
    "    #print(exp_a_aff, exp_b_aff, exp_a_unaf, exp_b_unaf)\n",
    "\n",
    "    chi_a_aff = ((a - exp_a_aff) ** 2) / exp_a_aff\n",
    "    chi_b_aff = ((b - exp_b_aff) ** 2) / exp_b_aff\n",
    "    chi_a_unaf = ((c - exp_a_unaf) ** 2) / exp_a_unaf\n",
    "    chi_b_unaf = ((d - exp_b_unaf) ** 2) / exp_b_unaf\n",
    "\n",
    "    x_2 = chi_a_aff + chi_b_aff + chi_a_unaf + chi_b_unaf\n",
    "\n",
    "    p_val = 1- stats.chi2.cdf(x_2, 1)\n",
    "    print('Chi-square test (1 df): ', x_2)  \n",
    "    print('P-value: ', p_val)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HCB181</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HCB182</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HCB183</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HCB184</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HCB185</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        0  1  2\n",
       "0  HCB181  1  1\n",
       "1  HCB182  1  1\n",
       "2  HCB183  1  1\n",
       "3  HCB184  1  1\n",
       "4  HCB185  1  1"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "pop_file = pd.read_csv(\"pop.phe\", sep='\\s+', header=None)\n",
    "pop_file.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# getting \n",
    "ids = pop_file[0]\n",
    "pop =  pop_file[2]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['rs6681049_1',\n",
       " 'rs6681049_2',\n",
       " 'rs4074137_1',\n",
       " 'rs4074137_2',\n",
       " 'rs7540009_1',\n",
       " 'rs7540009_2',\n",
       " 'rs1891905_1',\n",
       " 'rs1891905_2',\n",
       " 'rs9729550_1',\n",
       " 'rs9729550_2']"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_ids = pd.read_csv(\"hapmap1.map\", sep='\\s+', header=None)[1]\n",
    "snp_ids_diploid = [(x+\"_1\", x+\"_2\") for x in list(snp_ids)]\n",
    "snp_ids_diploid = list(itertools.chain(*snp_ids_diploid))\n",
    "snp_ids_diploid[0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "genotype = pd.read_csv(\"hapmap1.ped\", sep=\"\\s+\", header=None)\n",
    "genotype['populations'] = pop\n",
    "genotype.iloc[:10, :10]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "col_names = [\"individual\", \"X\", \"X\", \"X\", \"X\", \"phenotype\"] + snp_ids_diploid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>individual</th>\n",
       "      <th>X</th>\n",
       "      <th>X</th>\n",
       "      <th>X</th>\n",
       "      <th>X</th>\n",
       "      <th>phenotype</th>\n",
       "      <th>rs6681049_1</th>\n",
       "      <th>rs6681049_2</th>\n",
       "      <th>rs4074137_1</th>\n",
       "      <th>rs4074137_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HCB181</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HCB182</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HCB183</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HCB184</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HCB185</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HCB186</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HCB187</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HCB188</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HCB189</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HCB190</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  individual  X  X  X  X  phenotype  rs6681049_1  rs6681049_2  rs4074137_1  \\\n",
       "0     HCB181  1  0  0  1          1            2            2            2   \n",
       "1     HCB182  1  0  0  1          1            2            2            1   \n",
       "2     HCB183  1  0  0  1          2            2            2            1   \n",
       "3     HCB184  1  0  0  1          1            2            2            1   \n",
       "4     HCB185  1  0  0  1          1            2            2            1   \n",
       "5     HCB186  1  0  0  1          1            2            2            2   \n",
       "6     HCB187  1  0  0  1          1            2            2            2   \n",
       "7     HCB188  1  0  0  1          1            2            2            1   \n",
       "8     HCB189  1  0  0  1          1            2            2            2   \n",
       "9     HCB190  1  0  0  1          1            2            2            2   \n",
       "\n",
       "   rs4074137_2  \n",
       "0            2  \n",
       "1            2  \n",
       "2            2  \n",
       "3            2  \n",
       "4            2  \n",
       "5            2  \n",
       "6            2  \n",
       "7            2  \n",
       "8            2  \n",
       "9            2  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotype.columns = col_names\n",
    "genotype.iloc[:10, :10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#### Heer resort the column headers so that:\n",
    "# 1- Individual is your index\n",
    "# 2- pop and phenotype are first columns, followed by snp_ids\n",
    "# 3- remove the x columns if there are not needed or rename them to their proper names if they are needed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>individual</th>\n",
       "      <th>X</th>\n",
       "      <th>X</th>\n",
       "      <th>X</th>\n",
       "      <th>X</th>\n",
       "      <th>phenotype</th>\n",
       "      <th>rs6681049_1</th>\n",
       "      <th>rs6681049_2</th>\n",
       "      <th>rs4074137_1</th>\n",
       "      <th>rs4074137_2</th>\n",
       "      <th>...</th>\n",
       "      <th>rs2269380_2</th>\n",
       "      <th>rs6151412_1</th>\n",
       "      <th>rs6151412_2</th>\n",
       "      <th>rs11912064_1</th>\n",
       "      <th>rs11912064_2</th>\n",
       "      <th>rs1001469_1</th>\n",
       "      <th>rs1001469_2</th>\n",
       "      <th>rs756638_1</th>\n",
       "      <th>rs756638_2</th>\n",
       "      <th>populations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HCB181</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HCB182</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HCB184</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HCB185</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HCB186</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 167075 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  individual  X  X  X  X  phenotype  rs6681049_1  rs6681049_2  rs4074137_1  \\\n",
       "0     HCB181  1  0  0  1          1            2            2            2   \n",
       "1     HCB182  1  0  0  1          1            2            2            1   \n",
       "3     HCB184  1  0  0  1          1            2            2            1   \n",
       "4     HCB185  1  0  0  1          1            2            2            1   \n",
       "5     HCB186  1  0  0  1          1            2            2            2   \n",
       "\n",
       "   rs4074137_2     ...       rs2269380_2  rs6151412_1  rs6151412_2  \\\n",
       "0            2     ...                 2            2            2   \n",
       "1            2     ...                 2            2            2   \n",
       "3            2     ...                 2            2            2   \n",
       "4            2     ...                 2            2            2   \n",
       "5            2     ...                 2            2            2   \n",
       "\n",
       "   rs11912064_1  rs11912064_2  rs1001469_1  rs1001469_2  rs756638_1  \\\n",
       "0             2             2            1            2           2   \n",
       "1             2             2            1            2           2   \n",
       "3             2             2            1            1           1   \n",
       "4             2             2            1            2           2   \n",
       "5             2             2            1            2           2   \n",
       "\n",
       "   rs756638_2  populations  \n",
       "0           2            1  \n",
       "1           2            1  \n",
       "3           2            1  \n",
       "4           2            1  \n",
       "5           2            1  \n",
       "\n",
       "[5 rows x 167075 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "df_pheno1 = genotype.loc[genotype['phenotype'] == 1]\n",
    "df_pheno2 = genotype.loc[genotype['phenotype'] == 2]\n",
    "df_pheno1.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>individual</th>\n",
       "      <th>X</th>\n",
       "      <th>X</th>\n",
       "      <th>X</th>\n",
       "      <th>X</th>\n",
       "      <th>phenotype</th>\n",
       "      <th>rs6681049_1</th>\n",
       "      <th>rs6681049_2</th>\n",
       "      <th>rs4074137_1</th>\n",
       "      <th>rs4074137_2</th>\n",
       "      <th>...</th>\n",
       "      <th>rs2269380_2</th>\n",
       "      <th>rs6151412_1</th>\n",
       "      <th>rs6151412_2</th>\n",
       "      <th>rs11912064_1</th>\n",
       "      <th>rs11912064_2</th>\n",
       "      <th>rs1001469_1</th>\n",
       "      <th>rs1001469_2</th>\n",
       "      <th>rs756638_1</th>\n",
       "      <th>rs756638_2</th>\n",
       "      <th>populations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HCB181</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HCB182</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HCB183</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HCB184</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HCB185</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 167075 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  individual  X  X  X  X  phenotype  rs6681049_1  rs6681049_2  rs4074137_1  \\\n",
       "0     HCB181  1  0  0  1          1            2            2            2   \n",
       "1     HCB182  1  0  0  1          1            2            2            1   \n",
       "2     HCB183  1  0  0  1          2            2            2            1   \n",
       "3     HCB184  1  0  0  1          1            2            2            1   \n",
       "4     HCB185  1  0  0  1          1            2            2            1   \n",
       "\n",
       "   rs4074137_2     ...       rs2269380_2  rs6151412_1  rs6151412_2  \\\n",
       "0            2     ...                 2            2            2   \n",
       "1            2     ...                 2            2            2   \n",
       "2            2     ...                 1            2            2   \n",
       "3            2     ...                 2            2            2   \n",
       "4            2     ...                 2            2            2   \n",
       "\n",
       "   rs11912064_1  rs11912064_2  rs1001469_1  rs1001469_2  rs756638_1  \\\n",
       "0             2             2            1            2           2   \n",
       "1             2             2            1            2           2   \n",
       "2             2             2            2            2           1   \n",
       "3             2             2            1            1           1   \n",
       "4             2             2            1            2           2   \n",
       "\n",
       "   rs756638_2  populations  \n",
       "0           2            1  \n",
       "1           2            1  \n",
       "2           2            1  \n",
       "3           2            1  \n",
       "4           2            1  \n",
       "\n",
       "[5 rows x 167075 columns]"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_pop1 = genotype.loc[genotype['populations'] == 1]\n",
    "df_pop2 = genotype.loc[genotype['populations'] == 2]\n",
    "df_pop1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({'11': 8, '12': 22, '22': 59})"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_id = 'rs6681049'\n",
    "# maf_pop1, allele_pop1= get_al_freq('rs6681049', df_pop1)\n",
    "\n",
    "snp_genotype = genotype[snp_id+\"_1\"].astype('str') +  + genotype[snp_id+\"_2\"].astype('str')\n",
    "\n",
    "# usee pands counts number of occurrences\n",
    "Counter(list(snp_genotype)) / sum(Counter (list(snp_genotype)).vales())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_al_freq(snp_id, pop):\n",
    "    aa = 0\n",
    "    bb = 0\n",
    "    ab = 0\n",
    "    //total = 2*len(pop[[snp_id]])\n",
    "    \n",
    "    \n",
    "    \n",
    "    al_1 = 1\n",
    "    al_2 = 2\n",
    "\n",
    "    for i in range(len(pop[[snp_id]])):\n",
    "        \n",
    "        if pop[[snp_id]].iloc[i].iloc[0] == '2' and pop[[snp_id]].iloc[i].iloc[1] == '2':\n",
    "            aa += 1\n",
    "        elif pop[[snp_id]].iloc[i].iloc[0] == '1' and pop[[snp_id]].iloc[i].iloc[1] == '1':\n",
    "            bb += 1\n",
    "        else:\n",
    "            ab += 1\n",
    "\n",
    "    obs_a = 2*aa + ab\n",
    "    obs_b = 2*bb + ab\n",
    "\n",
    "    f1 = (2*aa + ab) / total\n",
    "    f2 = (2*bb + ab) / total\n",
    "\n",
    "    return which_minor_al(f1, f2, al_1, al_2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "rs6704013: # of missing:  2  frequency:  0.02247191011235955\n",
      "rs307347: # of missing:  12  frequency:  0.1348314606741573\n",
      "rs9439440: # of missing:  2  frequency:  0.02247191011235955\n",
      "\n",
      "SNP ID: rs6681049\n",
      "MAF Population 1:  0.23333333333333334 Minor allele:  2\n",
      "MAF Population 2:  0.19318181818181818 Minor allele:  2\n",
      "Frequency in unaffected:  0.26666666666666666 Allele:  2\n",
      "Frequency in affected:  0.1590909090909091 Allele:  2\n",
      "\n",
      "Basic Association Test rs6681049: \n",
      "Chi-square test (1 df):  3.066637047163362\n",
      "P-value:  0.07991463136564891\n",
      "\n",
      "Basic Association Test rs9585021: \n",
      "Chi-square test (1 df):  20.27001965692306\n",
      "P-value:  6.7246097376072456e-06\n",
      "\n",
      "Basic Association Test rs2222162: \n",
      "Chi-square test (1 df):  20.514418590898156\n",
      "P-value:  5.918374757829525e-06\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "maf_pop2, allele_pop2 = get_al_freq('rs6681049', df_pop2)\n",
    "phe_unaf_frq, allele_phe1 = get_al_freq('rs6681049', df_pheno1)\n",
    "phe_af_frq, allele_phe2 = get_al_freq('rs6681049', df_pheno2)\n",
    "\n",
    "# Missing statistics\n",
    "missing_nu, freq = get_missing('rs6704013', df)\n",
    "print('rs6704013: # of missing: ', missing_nu, ' frequency: ', freq)\n",
    "\n",
    "missing_nu, freq = get_missing('rs307347', df)\n",
    "print('rs307347: # of missing: ', missing_nu, ' frequency: ', freq)\n",
    "\n",
    "missing_nu, freq = get_missing('rs9439440', df)\n",
    "print('rs9439440: # of missing: ', missing_nu, ' frequency: ', freq)\n",
    "\n",
    "# Minor allele statistics\n",
    "print()\n",
    "print('SNP ID: rs6681049')\n",
    "print('MAF Population 1: ', maf_pop1, \"Minor allele: \", allele_pop1)\n",
    "print('MAF Population 2: ', maf_pop2, \"Minor allele: \", allele_pop2)\n",
    "print('Frequency in unaffected: ', phe_unaf_frq, \"Allele: \", allele_phe1)\n",
    "print('Frequency in affected: ', phe_af_frq, \"Allele: \", allele_phe2)\n",
    "\n",
    "# Basic Association Test\n",
    "print()\n",
    "df_snp1 = df[['Pheno', 'rs6681049']]\n",
    "df_snp2 = df[['Pheno', 'rs9585021']]\n",
    "df_snp3 = df[['Pheno', 'rs2222162']]\n",
    "\n",
    "print('Basic Association Test rs6681049: ')\n",
    "chi_square_test(df_snp1)\n",
    "\n",
    "print()\n",
    "print('Basic Association Test rs9585021: ' )\n",
    "chi_square_test(df_snp2)\n",
    "\n",
    "print()\n",
    "print('Basic Association Test rs2222162: ')\n",
    "chi_square_test(df_snp3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note: The test for rs6958021 does not match the result in PLINK but the other SNPs do."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
