{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import itertools\n",
    "from scipy import stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_gt(filename):\n",
    "    with open(filename, \"r\") as fileHandler:\n",
    "        lines = fileHandler.readlines()\n",
    "    return lines\n",
    "\n",
    "def get_pop(filename):\n",
    "    phenotypes = []\n",
    "    ids = []\n",
    "    with open(filename, \"r\") as fileHandler:\n",
    "        lines = fileHandler.readlines()\n",
    "    for i in range(len(lines)):\n",
    "        ids.append(lines[i].split()[0])\n",
    "        phenotypes.append(lines[i].split()[2])\n",
    "    return ids, phenotypes\n",
    "\n",
    "def get_snp_ids(filename):\n",
    "    snp_ids = ['IND', 'X', 'X', 'X', 'X', 'Pheno']\n",
    "    with open(filename, \"r\") as fileHandler:\n",
    "        lines = fileHandler.readlines()\n",
    "    for i in range(len(lines)):\n",
    "        snp_ids.append(lines[i].split()[1])\n",
    "        snp_ids.append(lines[i].split()[1])\n",
    "    return snp_ids\n",
    "\n",
    "def get_ids(genotypes):\n",
    "    ids = []\n",
    "    for i in range(0, len(genotypes)):\n",
    "        ids.append(genotypes[i].split()[0])\n",
    "    return ids\n",
    "\n",
    "def get_data_frame(genotypes, snp_ids):\n",
    "    inds = []\n",
    "    for i in range(0, len(genotypes)):\n",
    "        inds.append(genotypes[i].split())\n",
    "    df = pd.DataFrame(inds)\n",
    "    df.columns = snp_ids\n",
    "    return df\n",
    "\n",
    "def which_minor_al(frq1, frq2, a1, a2):\n",
    "    if frq1 < frq2:\n",
    "        return frq1, a1\n",
    "    else:\n",
    "        return frq2, a2\n",
    "\n",
    "def get_al_freq(snp_id, pop):\n",
    "    aa = 0\n",
    "    bb = 0\n",
    "    ab = 0\n",
    "    total = 2*len(pop[[snp_id]])\n",
    "    al_1 = \"1\"\n",
    "    al_2 = \"2\"\n",
    "\n",
    "    for i in range(len(pop[[snp_id]])):\n",
    "        \n",
    "        if pop[[snp_id]].iloc[i].iloc[0] == '2' and pop[[snp_id]].iloc[i].iloc[1] == '2':\n",
    "            aa += 1\n",
    "        elif pop[[snp_id]].iloc[i].iloc[0] == '1' and pop[[snp_id]].iloc[i].iloc[1] == '1':\n",
    "            bb += 1\n",
    "        else:\n",
    "            ab += 1\n",
    "\n",
    "    obs_a = 2*aa + ab\n",
    "    obs_b = 2*bb + ab\n",
    "\n",
    "    f1 = (2*aa + ab) / total\n",
    "    f2 = (2*bb + ab) / total\n",
    "\n",
    "    return which_minor_al(f1, f2, al_1, al_2)\n",
    "\n",
    "def get_missing(snp_id, df):\n",
    "    missing_nu = 0\n",
    "    for i in range(len(df[[snp_id]])):\n",
    "        #if df[[snp_id]].iloc[i].all() == '0':\n",
    "         #   missing_nu += 2\n",
    "        if df[[snp_id]].iloc[i].any() == '0':\n",
    "            missing_nu += 1\n",
    "    return missing_nu, missing_nu / len(df[[snp_id]])\n",
    "\n",
    "def missing_snps(person, nu_snps):\n",
    "    missing = 0\n",
    "    for i in range(nu_snps-6):\n",
    "        print(i+6)\n",
    "        if person[i+6] == '0' or person[i+7] == '0':\n",
    "            missing += 1\n",
    "        i += 2\n",
    "    return missing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def chi_square_test(pop):\n",
    "    aa_aff = 0\n",
    "    bb_aff = 0\n",
    "    ab_aff = 0\n",
    "\n",
    "    aa_unaf = 0\n",
    "    bb_unaf = 0\n",
    "    ab_unaf = 0\n",
    "\n",
    "    AFFECTED = '2'\n",
    "    CONTROL = '1'\n",
    "\n",
    "    for i in range(len(pop)):\n",
    "        if pop.iloc[i].iloc[1] == '2' and pop.iloc[i].iloc[2] == '2' and pop.iloc[i].iloc[0] == AFFECTED:\n",
    "            aa_aff += 1\n",
    "        elif pop.iloc[i].iloc[1] == '1' and pop.iloc[i].iloc[2] == '1' and pop.iloc[i].iloc[0] == AFFECTED:\n",
    "            bb_aff += 1\n",
    "        elif pop.iloc[i].iloc[0] == AFFECTED:\n",
    "            ab_aff += 1\n",
    "        \n",
    "        if pop.iloc[i].iloc[1] == '2' and pop.iloc[i].iloc[2] == '2' and pop.iloc[i].iloc[0] == CONTROL:\n",
    "            aa_unaf += 1\n",
    "        elif pop.iloc[i].iloc[1] == '1' and pop.iloc[i].iloc[2] == '1' and pop.iloc[i].iloc[0] == CONTROL:\n",
    "            bb_unaf += 1\n",
    "        elif pop.iloc[i].iloc[0] == CONTROL:\n",
    "            ab_unaf += 1\n",
    "\n",
    "    a = 2*aa_aff + ab_aff\n",
    "    b = 2*bb_aff + ab_aff\n",
    "    c = 2*aa_unaf + ab_unaf\n",
    "    d = 2*bb_unaf + ab_unaf\n",
    "\n",
    "    total = a + b + c + d\n",
    "\n",
    "    #print (a, b, c, d)\n",
    "    a_aff = a / total\n",
    "    b_aff = b / total\n",
    "    a_unaf = c / total\n",
    "    b_unaf = d / total\n",
    "    #print(a_aff, b_aff, a_unaf, b_unaf)\n",
    "\n",
    "    p_aff = a_aff + b_aff\n",
    "    p_unaf = a_unaf + b_unaf\n",
    "    p_a = a_aff + a_unaf\n",
    "    p_b = b_aff + b_unaf\n",
    "    #print(p_aff, p_unaf, p_a, p_b)\n",
    "\n",
    "    p_a_aff = p_aff * p_a\n",
    "    p_b_aff = p_aff * p_b\n",
    "    p_a_unaf = p_unaf * p_a\n",
    "    p_b_unaf = p_unaf * p_b\n",
    "    #print(p_a_aff, p_b_aff, p_a_unaf, p_b_unaf)\n",
    "\n",
    "    exp_a_aff = total * p_a_aff\n",
    "    exp_b_aff = total * p_b_aff\n",
    "    exp_a_unaf = total * p_a_unaf\n",
    "    exp_b_unaf = total * p_b_unaf\n",
    "    #print(exp_a_aff, exp_b_aff, exp_a_unaf, exp_b_unaf)\n",
    "\n",
    "    chi_a_aff = ((a - exp_a_aff) ** 2) / exp_a_aff\n",
    "    chi_b_aff = ((b - exp_b_aff) ** 2) / exp_b_aff\n",
    "    chi_a_unaf = ((c - exp_a_unaf) ** 2) / exp_a_unaf\n",
    "    chi_b_unaf = ((d - exp_b_unaf) ** 2) / exp_b_unaf\n",
    "\n",
    "    x_2 = chi_a_aff + chi_b_aff + chi_a_unaf + chi_b_unaf\n",
    "\n",
    "    p_val = 1- stats.chi2.cdf(x_2, 1)\n",
    "    print('Chi-square test (1 df): ', x_2)  \n",
    "    print('P-value: ', p_val)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read in the categorical phenotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HCB181</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HCB182</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HCB183</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HCB184</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HCB185</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        0  1  2\n",
       "0  HCB181  1  1\n",
       "1  HCB182  1  1\n",
       "2  HCB183  1  1\n",
       "3  HCB184  1  1\n",
       "4  HCB185  1  1"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pop_file = pd.read_csv(\"pop.phe\", sep='\\s+', header=None)\n",
    "pop_file.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# getting \n",
    "# IDs are the first column\n",
    "# The phenotype encodings are the second column.\n",
    "ids = pop_file[0]\n",
    "pop =  pop_file[2]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Duplicate SNP IDs\n",
    "Since each SNP has to values (one for each chromosome copy) we need to provide each SNP ID with a extra character so we know what SNP we are looking at. Here we add _1 and _2 to each SNP ID."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['rs6681049_1',\n",
       " 'rs6681049_2',\n",
       " 'rs4074137_1',\n",
       " 'rs4074137_2',\n",
       " 'rs7540009_1',\n",
       " 'rs7540009_2',\n",
       " 'rs1891905_1',\n",
       " 'rs1891905_2',\n",
       " 'rs9729550_1',\n",
       " 'rs9729550_2']"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_ids = pd.read_csv(\"hapmap1.map\", sep='\\s+', header=None)[1]\n",
    "snp_ids_diploid = [(x+\"_1\", x+\"_2\") for x in list(snp_ids)]\n",
    "snp_ids_diploid = list(itertools.chain(*snp_ids_diploid))\n",
    "snp_ids_diploid[0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HCB181</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HCB182</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HCB183</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HCB184</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HCB185</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HCB186</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HCB187</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HCB188</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HCB189</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HCB190</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        0  1  2  3  4  5  6  7  8  9\n",
       "0  HCB181  1  0  0  1  1  2  2  2  2\n",
       "1  HCB182  1  0  0  1  1  2  2  1  2\n",
       "2  HCB183  1  0  0  1  2  2  2  1  2\n",
       "3  HCB184  1  0  0  1  1  2  2  1  2\n",
       "4  HCB185  1  0  0  1  1  2  2  1  2\n",
       "5  HCB186  1  0  0  1  1  2  2  2  2\n",
       "6  HCB187  1  0  0  1  1  2  2  2  2\n",
       "7  HCB188  1  0  0  1  1  2  2  1  2\n",
       "8  HCB189  1  0  0  1  1  2  2  2  2\n",
       "9  HCB190  1  0  0  1  1  2  2  2  2"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotype = pd.read_csv(\"hapmap1.ped\", sep=\"\\s+\", header=None)\n",
    "genotype['populations'] = pop\n",
    "genotype.iloc[:10, :10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We now need to adjust the column names to include the individual IDS, phenotypes, and the SNP IDs we just made."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "col_names = [\"individual\", \"X1\", \"X2\", \"X3\", \"X4\", \"phenotype\"] + snp_ids_diploid + ['populations']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>individual</th>\n",
       "      <th>X1</th>\n",
       "      <th>X2</th>\n",
       "      <th>X3</th>\n",
       "      <th>X4</th>\n",
       "      <th>phenotype</th>\n",
       "      <th>rs6681049_1</th>\n",
       "      <th>rs6681049_2</th>\n",
       "      <th>rs4074137_1</th>\n",
       "      <th>rs4074137_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HCB181</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HCB182</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HCB183</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HCB184</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HCB185</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HCB186</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HCB187</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HCB188</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HCB189</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HCB190</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  individual  X1  X2  X3  X4  phenotype  rs6681049_1  rs6681049_2  \\\n",
       "0     HCB181   1   0   0   1          1            2            2   \n",
       "1     HCB182   1   0   0   1          1            2            2   \n",
       "2     HCB183   1   0   0   1          2            2            2   \n",
       "3     HCB184   1   0   0   1          1            2            2   \n",
       "4     HCB185   1   0   0   1          1            2            2   \n",
       "5     HCB186   1   0   0   1          1            2            2   \n",
       "6     HCB187   1   0   0   1          1            2            2   \n",
       "7     HCB188   1   0   0   1          1            2            2   \n",
       "8     HCB189   1   0   0   1          1            2            2   \n",
       "9     HCB190   1   0   0   1          1            2            2   \n",
       "\n",
       "   rs4074137_1  rs4074137_2  \n",
       "0            2            2  \n",
       "1            1            2  \n",
       "2            1            2  \n",
       "3            1            2  \n",
       "4            1            2  \n",
       "5            2            2  \n",
       "6            2            2  \n",
       "7            1            2  \n",
       "8            2            2  \n",
       "9            2            2  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotype.columns = col_names\n",
    "genotype.iloc[:10, :10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Rearrange the columns and set the index of the dataframe to be the individual IDs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>populations</th>\n",
       "      <th>phenotype</th>\n",
       "      <th>X1</th>\n",
       "      <th>X2</th>\n",
       "      <th>X3</th>\n",
       "      <th>X4</th>\n",
       "      <th>rs6681049_1</th>\n",
       "      <th>rs6681049_2</th>\n",
       "      <th>rs4074137_1</th>\n",
       "      <th>rs4074137_2</th>\n",
       "      <th>...</th>\n",
       "      <th>rs2269380_1</th>\n",
       "      <th>rs2269380_2</th>\n",
       "      <th>rs6151412_1</th>\n",
       "      <th>rs6151412_2</th>\n",
       "      <th>rs11912064_1</th>\n",
       "      <th>rs11912064_2</th>\n",
       "      <th>rs1001469_1</th>\n",
       "      <th>rs1001469_2</th>\n",
       "      <th>rs756638_1</th>\n",
       "      <th>rs756638_2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>individual</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HCB181</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HCB182</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HCB183</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HCB184</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HCB185</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 167074 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            populations  phenotype  X1  X2  X3  X4  rs6681049_1  rs6681049_2  \\\n",
       "individual                                                                     \n",
       "HCB181                1          1   1   0   0   1            2            2   \n",
       "HCB182                1          1   1   0   0   1            2            2   \n",
       "HCB183                1          2   1   0   0   1            2            2   \n",
       "HCB184                1          1   1   0   0   1            2            2   \n",
       "HCB185                1          1   1   0   0   1            2            2   \n",
       "\n",
       "            rs4074137_1  rs4074137_2  ...  rs2269380_1  rs2269380_2  \\\n",
       "individual                            ...                             \n",
       "HCB181                2            2  ...            2            2   \n",
       "HCB182                1            2  ...            2            2   \n",
       "HCB183                1            2  ...            1            1   \n",
       "HCB184                1            2  ...            1            2   \n",
       "HCB185                1            2  ...            1            2   \n",
       "\n",
       "            rs6151412_1  rs6151412_2  rs11912064_1  rs11912064_2  rs1001469_1  \\\n",
       "individual                                                                      \n",
       "HCB181                2            2             2             2            1   \n",
       "HCB182                2            2             2             2            1   \n",
       "HCB183                2            2             2             2            2   \n",
       "HCB184                2            2             2             2            1   \n",
       "HCB185                2            2             2             2            1   \n",
       "\n",
       "            rs1001469_2  rs756638_1  rs756638_2  \n",
       "individual                                       \n",
       "HCB181                2           2           2  \n",
       "HCB182                2           2           2  \n",
       "HCB183                2           1           2  \n",
       "HCB184                1           1           2  \n",
       "HCB185                2           2           2  \n",
       "\n",
       "[5 rows x 167074 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#### Here re-sort the column headers so that:\n",
    "# 1- Individual is your index\n",
    "# 2- pop and phenotype are first columns, followed by snp_ids\n",
    "# 3- remove the x columns if there are not needed or rename them to their proper names if they are needed.\n",
    "re_col_names = ['individual', 'populations', 'phenotype', 'X1', 'X2', 'X3', 'X4'] + snp_ids_diploid\n",
    "genotype = genotype[re_col_names]\n",
    "genotype = genotype.set_index('individual')\n",
    "genotype.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The above restructring allows us to access individual data by using the individual's ID along with the attribute we want. For example, if we want to access individual HCB181 and look at there phenotype, population, and allele for SNP rs6681049 we could do the following:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "populations    1\n",
       "phenotype      1\n",
       "rs6681049_1    2\n",
       "rs6681049_2    2\n",
       "Name: HCB181, dtype: int64"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotype.loc['HCB181'][['populations', 'phenotype', 'rs6681049_1', 'rs6681049_2']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we can see that individual HCB181 belongs to population 1, has phenotype code 1, and the SNP has a homozygous genotype for the second variant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import Counter"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can now filter by each population and obtain SNP frequencies. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter by population 1 and 2\n",
    "genotypes_pop_1 = genotype[genotype['populations'] == 1]\n",
    "genotypes_pop_2 = genotype[genotype['populations'] == 2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define SNP ID we want to look at\n",
    "snp_id = 'rs6681049'\n",
    "\n",
    "# Grab SNP values within each population\n",
    "snp_genotype_pop_1 = genotypes_pop_1[snp_id+\"_1\"].astype('str') +  + genotypes_pop_1[snp_id+\"_2\"].astype('str')\n",
    "snp_genotype_pop_2 = genotypes_pop_2[snp_id+\"_1\"].astype('str') +  + genotypes_pop_2[snp_id+\"_2\"].astype('str')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Count the occurrences of each allele and obtain frequencies\n",
    "genotype_counts_pop_1 = Counter(list(snp_genotype_pop_1))\n",
    "genotype_counts_pop_2 = Counter(list(snp_genotype_pop_2))\n",
    "\n",
    "count_aa_pop_1 = genotype_counts_pop_1['22']\n",
    "count_bb_pop_1 = genotype_counts_pop_1['11']\n",
    "count_ab_pop_1 = genotype_counts_pop_1['12']\n",
    "\n",
    "count_aa_pop_2 = genotype_counts_pop_2['22']\n",
    "count_bb_pop_2 = genotype_counts_pop_2['11']\n",
    "count_ab_pop_2 = genotype_counts_pop_2['12']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Allele 1 freq:  0.7666666666666667\n",
      "Allele 2 freq:  0.23333333333333334\n"
     ]
    }
   ],
   "source": [
    "# We then define the number of occurrences by twice the total since we are counting by pairs of alleles\n",
    "total = count_aa_pop_1 + count_bb_pop_1 + count_ab_pop_1\n",
    "A1_frequency = (2*count_aa_pop_1 + count_ab_pop_1) / (2*total)\n",
    "A2_frequency = (2*count_bb_pop_1 + count_ab_pop_1) / (2*total)\n",
    "print('Population 1 Allele 1 freq: ', A1_frequency)\n",
    "print('Population 1 Allele 2 freq: ', A2_frequency)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Population 2 Allele 1 freq:  0.7888888888888889\n",
      "Population 2 Allele 2 freq:  0.18888888888888888\n"
     ]
    }
   ],
   "source": [
    "A1_frequency = (2*count_aa_pop_2 + count_ab_pop_2) / (2*total)\n",
    "A2_frequency = (2*count_bb_pop_2 + count_ab_pop_2) / (2*total)\n",
    "print('Population 2 Allele 1 freq: ', A1_frequency)\n",
    "print('Population 2 Allele 2 freq: ', A2_frequency)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When considering the frequency of SNPs we only consider the minor allele frequency which is the frequency of the allele that occurs less. Thus, in population 1 we see a minor allele frequency of 0.233 and for population 2 we see a minor allele frequency of 0.188 for this particular SNP."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MAF Unaffected:  0.26666666666666666\n",
      "MAF Affected:  0.1590909090909091\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Power_divergenceResult(statistic=1.0172661229830218, pvalue=0.7970741635167385)"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Perform Basic Association Test\n",
    "\n",
    "# Filter by different phenotypes for specific SNP\n",
    "GT_uf = genotype[genotype['phenotype'] == 1][[snp_id+'_1', snp_id+'_2']]\n",
    "GT_af = genotype[genotype['phenotype'] == 2][[snp_id+'_1', snp_id+'_2']]\n",
    "\n",
    "snp_GT_uf = GT_uf[snp_id+\"_1\"].astype('str') +  + GT_uf[snp_id+\"_2\"].astype('str')\n",
    "snp_GT_af = GT_af[snp_id+\"_1\"].astype('str') +  + GT_af[snp_id+\"_2\"].astype('str')\n",
    "\n",
    "# get MAF for affected and unaffected\n",
    "GT_counts_uf = Counter(list(snp_GT_uf))\n",
    "GT_counts_af = Counter(list(snp_GT_af))\n",
    "\n",
    "count_aa_GT_uf = GT_counts_uf['22']\n",
    "count_bb_GT_uf = GT_counts_uf['11']\n",
    "count_ab_GT_uf = GT_counts_uf['12']\n",
    "uf_total = 2*(count_aa_GT_uf + count_bb_GT_uf + count_ab_GT_uf)\n",
    "\n",
    "count_aa_GT_af = GT_counts_af['22']\n",
    "count_bb_GT_af = GT_counts_af['11']\n",
    "count_ab_GT_af = GT_counts_af['12']\n",
    "af_total = 2*(count_aa_GT_af + count_bb_GT_af + count_ab_GT_af)\n",
    "\n",
    "a_uf = (2*count_aa_GT_uf + count_ab_GT_uf) / uf_total\n",
    "b_uf = (2*count_bb_GT_uf + count_ab_GT_uf) / uf_total\n",
    "\n",
    "a_af = (2*count_aa_GT_af + count_ab_GT_af) / af_total\n",
    "b_af = (2*count_bb_GT_af + count_ab_GT_af) / af_total\n",
    "\n",
    "print('MAF Unaffected: ', b_uf)\n",
    "print('MAF Affected: ', b_af)\n",
    "\n",
    "# perform test\n",
    "p_aff = a_af + b_af\n",
    "p_unaff = a_uf + b_uf\n",
    "p_a = a_af + a_uf\n",
    "p_b = b_af + b_uf\n",
    "\n",
    "exp_a_af = p_a * p_aff\n",
    "exp_a_uf = p_a * p_unaff\n",
    "exp_b_af = p_b * p_aff\n",
    "exp_b_uf = p_b * p_unaff\n",
    "\n",
    "a_count_uf = a_uf * uf_total\n",
    "a_count_af = a_af * af_total\n",
    "b_count_uf = b_uf * uf_total\n",
    "b_count_af = b_af * af_total\n",
    "\n",
    "\n",
    "\n",
    "stats.chisquare([a_uf, b_uf, a_af, b_af], [exp_a_uf, exp_b_uf, exp_a_af, exp_b_af])\n",
    "\n",
    "# get p-value\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "maf_pop2, allele_pop2 = get_al_freq('rs6681049', df_pop2)\n",
    "phe_unaf_frq, allele_phe1 = get_al_freq('rs6681049', df_pheno1)\n",
    "phe_af_frq, allele_phe2 = get_al_freq('rs6681049', df_pheno2)\n",
    "\n",
    "# Missing statistics\n",
    "missing_nu, freq = get_missing('rs6704013', df)\n",
    "print('rs6704013: # of missing: ', missing_nu, ' frequency: ', freq)\n",
    "\n",
    "missing_nu, freq = get_missing('rs307347', df)\n",
    "print('rs307347: # of missing: ', missing_nu, ' frequency: ', freq)\n",
    "\n",
    "missing_nu, freq = get_missing('rs9439440', df)\n",
    "print('rs9439440: # of missing: ', missing_nu, ' frequency: ', freq)\n",
    "\n",
    "# Minor allele statistics\n",
    "print()\n",
    "print('SNP ID: rs6681049')\n",
    "print('MAF Population 1: ', maf_pop1, \"Minor allele: \", allele_pop1)\n",
    "print('MAF Population 2: ', maf_pop2, \"Minor allele: \", allele_pop2)\n",
    "print('Frequency in unaffected: ', phe_unaf_frq, \"Allele: \", allele_phe1)\n",
    "print('Frequency in affected: ', phe_af_frq, \"Allele: \", allele_phe2)\n",
    "\n",
    "# Basic Association Test\n",
    "print()\n",
    "df_snp1 = df[['Pheno', 'rs6681049']]\n",
    "df_snp2 = df[['Pheno', 'rs9585021']]\n",
    "df_snp3 = df[['Pheno', 'rs2222162']]\n",
    "\n",
    "print('Basic Association Test rs6681049: ')\n",
    "chi_square_test(df_snp1)\n",
    "\n",
    "print()\n",
    "print('Basic Association Test rs9585021: ' )\n",
    "chi_square_test(df_snp2)\n",
    "\n",
    "print()\n",
    "print('Basic Association Test rs2222162: ')\n",
    "chi_square_test(df_snp3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note: The test for rs6958021 does not match the result in PLINK but the other SNPs do."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
